---
title: "BRCA Gene Expression Analyses"
author: "Lizzy Ramsey"
date: "9/10/2022"
output: html_document
---

load packages (note: for first time using a package, it first must be installed. You can do this by running BiocManager::install("packagename") first before loading it library(packagename) )
```{r}
library(recount3) #for accessing data
library(dplyr) #great package for data wrangling
library(ggplot2) #package for plotting
library(ggfortify)
library(limma)
library(edgeR)
```

View projects in recount3 
```{r}
human_projects<- available_projects()
View(human_projects)
```
recount3 enables creating RangedSummarizedExperiment objects that contain expression quantitative data ([Figure 2](https://www.bioconductor.org/packages/devel/bioc/vignettes/recount3/inst/doc/recount3-quickstart.html) ). As a quick overview, some of the main terms are:  
__rse:__ a RangedSummarizedExperiment object from SummarizedExperiment (Morgan, Obenchain, Hester, and PagÃ¨s, 2019) that contains:  
__counts:__ a matrix with the expression feature data (either: gene, exon, or exon-exon junctions) and that can be accessed using assays(counts).  
__metadata:__ a table with information about the samples and quality metrics that can be accessed using colData(rse).  
__annotation:__ a table-like object with information about the expression features which can be annotation-specific (gene and exons) or annotation-agnostic (exon-exon junctions). This information can be accessed using rowRanges(rse).  
^^Taken from the [vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/recount3/inst/doc/recount3-quickstart.html) 
  
### TCGA
Create ranged summarized experiemnt (RSE) manual for TCGA BRCA data. Use the data listed for BRCA when looking at 'human_projects' from recount3  
```{r}
brca_data <- recount3::create_rse_manual(
    project = "BRCA",
    project_home = "data_sources/tcga",
    organism = "human",
    annotation = "gencode_v26",
    type = "gene")

```

Look at structure of rse object
```{r}
str(brca_data)
```

#### Explore metadata  
(take a good look at this - there's tons of metadata... we can include these things as covariates, decide to exclude samples based on certain criteria here, and/or make plots for some of these... there's morphology, tumor classification, patient ethnicity, age, days to death, treatments, clinical stage, pregnancies, and so much more)
```{r}
brca_metadata <- as.data.frame(colData(brca_data))
View(brca_metadata)
#what are all the available sample types?
unique(brca_metadata$tcga.cgc_sample_sample_type)
#what are the ages at diagnosis?
summary(brca_metadata$tcga.cgc_case_age_at_diagnosis) 
```


Let's focus on just Primary Tumors (metastastic ones are extra weird, and the solid tissue normal has been shown to not very 'normal' in previous studies)
```{r}
#filter for primary tumors
brca_primary_metadata <- dplyr::filter(brca_metadata, tcga.cgc_sample_sample_type == "Primary Tumor") #dplyr:: is just to say that I want to use the 'filter' function from the 'dplyr' package -- helps keep R from getting confused about using the 'filter' function from another package or base R 
```

Subset for patients with drug treatment data, count frequencies of each
```{r}
brca_drug_meta <- tidyr::drop_na(brca_primary_metadata, tcga.cgc_drug_therapy_drug_name)

#Frequency of treatments
trt_freq <- as.data.frame(table(brca_drug_meta$tcga.cgc_drug_therapy_drug_name))
View(trt_freq)
```

look at table--note the different counts for some drugs like "doxorubicin" "Doxorubicin""DOXORUBICIN" ... we need to make all lowercase 
```{r}
#change all drugs to completely lowercase
brca_drug_meta$tcga.cgc_drug_therapy_drug_name <- tolower(brca_drug_meta$tcga.cgc_drug_therapy_drug_name)

#Frequency of treatments
trt_freq <- as.data.frame(table(brca_drug_meta$tcga.cgc_drug_therapy_drug_name))
View(trt_freq)

```

Visualize most-used drugs and their frequencies
```{r}
top_trts <- dplyr::slice_max(trt_freq, order_by = Freq, n = 15)

ggplot(top_trts, aes(x = Var1, y = Freq, fill = Var1)) +  # Plot with values on top
  geom_bar(stat = "identity") + 
  geom_text(aes(label = Freq), vjust = 0) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5))

```

Subset data by patients treated with any of the top 10 most-frequent drugs
```{r}
#top 15 included some with just 9... went a little stricter, to top 10
top_trts <- dplyr::slice_max(trt_freq, order_by = Freq, n = 10)

#subset for patients treated with one of the top 10 most frequent drugs
brca_toptrt_meta <- dplyr::filter(brca_drug_meta, tcga.cgc_drug_therapy_drug_name %in% top_trts$Var1) 
#we can double check that those are the only treatments in this df 
unique(brca_toptrt_meta$tcga.cgc_drug_therapy_drug_name)

```

Stages of new subset (patients treated with a top 10 most-freq used drug)
```{r}
#what are the stages included?
unique(brca_toptrt_meta$tcga.xml_stage_event_pathologic_stage)
#Frequency of treatments
stage_freq <- as.data.frame(table(brca_drug_meta$tcga.xml_stage_event_pathologic_stage))
View(stage_freq)

#visualize
ggplot(stage_freq, aes(x = Var1, y = Freq, fill = Var1)) +  # Plot with values on top
  geom_bar(stat = "identity") + 
  geom_text(aes(label = Freq), vjust = 0) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5))

```

#### Counts data

View counts data from the RSE
```{r}
str(assays(brca_data))
```


Now generate Transcripts Per Million (TPM) from counts data
```{r}
#first, scale the raw counts
assay(brca_data, "counts") <- transform_counts(brca_data)
#calculate TPM
assays(brca_data)$TPM <- recount::getTPM(brca_data, length_var = "bp_length")
#view the counts data -- section for 'TPM' now
str(brca_data@assays@data@listData)
```


Data wrangling -- make counts into dataframe, then subset for the samples that are 'Primary Tumor' ones
```{r}
brca_tpm <- as.data.frame(brca_data@assays@data@listData$TPM)

brca_tpm <- dplyr::select(brca_tpm, matches(row.names(brca_primary_metadata)))
```


Save subsetted BRCA counts and metadata
```{r}
#please note-- to save to this filepath, you need a 'data' folder where ever your current working directory is (you can find out your working directory with getwd() ) so you can either change this filepath or create a 'data' folder in your current working directory
write.csv(brca_primary_metadata, "./data/brca_primarytumor_metadata.csv")

write.csv(brca_tpm, "./data/brca_primarytumor_tpm_expdata.csv")
```

##### Data Exploration
###### PCA of subset  
Pull counts for patients from top-treatment subset  
```{r}
toptrts_cts <- dplyr::select(brca_tpm, matches(row.names(brca_toptrt_meta)))

#take top 10,000 most variable genes -- PCA is a hefty computational process, and looks at variation, so we can grab the top most variant genes in the first place which should give a good representation for using all genes for PCA, but also is much easier to run on local computers
toptrts_cts_var <- apply(toptrts_cts, 1, IQR, na.rm = TRUE) #finds the interquartile range for each gene across samples; the 3rd qr - 1st qr. the greater the iqr, the more variability of  that gene across samples
varying_genes <- toptrts_cts_var[order(toptrts_cts_var, decreasing = TRUE)]
varying_genes <- names(varying_genes[1:10000])
#narrow down tpm to just top var genes
toptrts_cts_var_tpm <- toptrts_cts[rownames(toptrts_cts) %in% varying_genes,]

```


```{r}
#combining counts and metadata for pca
cts_t <- t(toptrts_cts_var_tpm)
meta_cts <- cbind(cts_t, brca_toptrt_meta)
toptrts_var_pca <- prcomp(meta_cts[1:10000], scale. = TRUE)
summary(toptrts_var_pca)
```


```{r}
# plotting 
toptrts_var_pca_plot <- autoplot(toptrts_var_pca,
						data = meta_cts,
						colour = 'tcga.gdc_cases.diagnoses.vital_status',  x = 11, y = 12)

toptrts_var_pca_plot
```

Save
```{r}
saveRDS(toptrts_var_pca, "./data/topfreq_drugs_pc.rds")

```

#### Features related to survival  
Area any of our explored features associated with survival?
```{r}
library(GGally)
#brca_toptrt_meta$tcga.cgc_case_age_at_diagnosis
GGally::ggpairs(data=brca_toptrt_meta, columns= c("tcga.gdc_cases.diagnoses.vital_status", "tcga.cgc_case_prior_diagnosis", "tcga.cgc_case_age_at_diagnosis"), title="Associations to Survival")

GGally::ggpairs(data=brca_toptrt_meta, columns= c("tcga.gdc_cases.diagnoses.vital_status", "tcga.gdc_cases.demographic.race", "tcga.cgc_case_age_at_diagnosis"), title="Associations to Survival")

GGally::ggpairs(data=brca_toptrt_meta, columns= c("tcga.gdc_cases.diagnoses.vital_status", "tcga.xml_stage_event_pathologic_stage", "tcga.cgc_drug_therapy_drug_name"), title="Associations to Survival")
```

```{r}
#fit_1 <- lm(tcga.gdc_cases.diagnoses.vital_status ~ tcga.cgc_case_age_at_diagnosis, data = brca_toptrt_meta)

```

##### For each drug of the top drugs, what's the breakdown of stages?  

Tamoxifen breakdown
```{r}
tamoxifen_set <- dplyr::filter(brca_toptrt_meta, tcga.cgc_drug_therapy_drug_name == "tamoxifen")

#what are the stages included?
unique(tamoxifen_set$tcga.xml_stage_event_pathologic_stage)
#Frequency of treatments
tamox_stage_freq <- as.data.frame(table(tamoxifen_set$tcga.xml_stage_event_pathologic_stage))
View(tamox_stage_freq)

#visualize
ggplot(tamox_stage_freq, aes(x = Var1, y = Freq, fill = Var1)) +  # Plot with values on top
  geom_bar(stat = "identity") + 
  geom_text(aes(label = Freq), vjust = 0) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5))

#age range and stats
summary(tamoxifen_set$tcga.cgc_case_age_at_diagnosis)

#days to death stats
summary(tamoxifen_set$tcga.xml_days_to_death)

#histology types
unique(tamoxifen_set$tcga.xml_histological_type)
```



### GTEx  
Looking at human_projects, find listing for GTEx breast data (just like done previously for TCGA BRCA)
```{r}
breast_data <- recount3::create_rse_manual(
    project = "BREAST",
    project_home = "data_sources/gtex",
    organism = "human",
    annotation = "gencode_v26",
    type = "gene")

```

Look at structure of rse object
```{r}
str(breast_data)
```

Explore metadata. We need to keep in mind the balance of data (for example, 'summary' of sex looks like it's almost half male/female, while the TCGA BRCA samples were almost all females. Also need to be aware of difference in reporting -- age for brca_metadata$tcga.cgc_case_age_at_diagnosis is exact age, and age for GTEx looks like it's in decade age ranges, i.e. 20-29, 30-39, etc )
```{r}
breast_metadata <- as.data.frame(colData(breast_data))
View(breast_metadata)
#all metadata columns
colnames(breast_metadata)
#what are all the available sample types?
unique(breast_metadata$gtex.age)
```

Analyze metadata
```{r}
#Frequency of ages
ctl_age_freq <- as.data.frame(table(breast_metadata$gtex.age))
View(ctl_age_freq)

ctl_rin_freq <- as.data.frame(table(breast_metadata$gtex.smrin))
View(ctl_rin_freq)

top_trts <- dplyr::slice_max(trt_freq, order_by = Freq, n = 15)

ggplot(top_trts, aes(x = Var1, y = Freq, fill = Var1)) +  # Plot with values on top
  geom_bar(stat = "identity") + 
  geom_text(aes(label = Freq), vjust = 0) +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.5))


```

View counts data from the RSE
```{r}
str(assays(breast_data))
```


Now generate Transcripts Per Million (TPM) from counts data
```{r}
#first, scale the raw counts
assay(breast_data, "counts") <- transform_counts(breast_data)
#calculate TPM
assays(breast_data)$TPM <- recount::getTPM(breast_data, length_var = "bp_length")
#view the counts data -- section for 'TPM' now
str(breast_data@assays@data@listData)
```


Data wrangling -- make counts into dataframe 
```{r}
breast_tpm <- as.data.frame(breast_data@assays@data@listData$TPM)

```


Save subsetted BRCA counts and metadata
```{r}
#please note-- to save to this filepath, you need a 'data' folder where ever your current working directory is (you can find out your working directory with getwd() ) so you can either change this filepath or create a 'data' folder in your current working directory
write.csv(breast_metadata, "./data/breast_healthy_metadata.csv")

write.csv(breast_tpm, "./data/breast_healthy_tpm_expdata.csv")
```

### Differential Expression Analysis  
LIMMA is a linear model with empirical bayes for finding differentially expressed genes.
[LIMMA user guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf)  
[Guide for creating a design matrix](https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html)
Random subset of first 10 samples for practice run through
```{r}
brca_test <- brca_tpm[,1:10]
ctl_test <- breast_tpm[,1:10]
cts_test <- cbind(brca_test, ctl_test)
```

Make new metadata, control vs tumor (can add in other features here to be used as covariates, i.e. age, race, etc.)
```{r}
lm_meta <- data.frame("Samples" = c(colnames(brca_test), colnames(ctl_test)), Condition = c(rep("Tumor", length(brca_test)), rep("Control", length(ctl_test))))

```

Make a design matrix
```{r}
des <- model.matrix(~lm_meta$Condition)

```

Filtering out genes with zero variance
```{r}
#makes object for convenient use
dge  <- edgeR::DGEList(counts=cts_test)
#remove genes consistently zero or very low
keep  <- edgeR::filterByExpr(dge , design = des)
dge  <- dge[keep ,,keep.lib.sizes=FALSE]

```

"When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend. In this approach, the voom transformation is applied to the normalized and filtered DGEList object:"
```{r}
v <- limma::voom(dge, des, plot=TRUE)
#The voom transformation uses the experiment design matrix, and produces an EList object.
#It is also possible to give a matrix of counts directly to voom without TMM normalization, by
v2 <- limma::voom(cts_test, des, plot=TRUE)
#If the data are very noisy, one can apply the same between-array normalization methods as would be used for microarrays, for example:
v3 <- limma::voom(cts_test, des, plot=TRUE, normalize="quantile")

# do the split on v

fractionTraining   <- 0.80
fractionTest       <- 0.20

# Compute sample sizes.
sampleSizeTraining   <- floor(fractionTraining   * nrow(v))
sampleSizeTest       <- floor(fractionTest       * nrow(v))

# Create the randomly-sampled indices for the dataframe. Use setdiff() to
# avoid overlapping subsets of indices.
indicesTraining    <- sort(sample(seq_len(nrow(v)), size=sampleSizeTraining))
indicesNotTraining <- setdiff(seq_len(nrow(v)), indicesTraining)
indicesTest  <- sort(sample(indicesNotTraining, size=sampleSizeTest))

# Finally, output the two dataframes for training, validation and test.
vTraining   <- v[indicesTraining, ]
vTest       <- v[indicesTest, ]

#After this, the usual limma pipelines for differential expression can be applied, for example:
fit <- limma::lmFit(v, des)
fit <- limma::eBayes(fit)
limma::topTable(fit, coef=ncol(des))
```
Filtering results
```{r}
#absolute log2fc > 2
#fit  <- limma::treat(fit , fc = 1.2, trend=TRUE)

#compute BH p-adjusted values, only values meeting 0.05 cutoff, avg expression, and all in nice table like top 10
#lm_res  <- topTreat(fit , coef=ncol(des ), sort.by = "logFC", number = nrow(fit [["coefficients"]]), p.value = 0.05)


```
